doc/control-mechanisms/reporting.md

```diff
@@ -0,0 +1,81 @@
+---
+title: "Reporting Konzept"
+---
+
+## Ziel
+
+Das Ziel dieses Dokuments ist es, die Konzepte zum **Reporting** zentral zu sammeln.
+**Reporting** bezeichnet den Prozess des Erfassens, Bewertens und Speicherns einer Antwort auf
+eine Frage im Dungeon-Kontext.
+
+## Beteiligte Komponenten
+
+**Task-Definition**
+
+Die Task-Definition definiert die Aufgabe, Beispiel:
+
+```
+single_choice_task my_task {
+  answers: ["a", "b", "c"],  // die Antwortmöglichkeiten
+  correct_answer: answers.0, // die korrekte Antwortmöglichkeit
+  fn_score: score            // die Scoring-Funktion/Bewertungsfunktion
+}
+```
+
+**Scoring-Funktion**
+
+Die **Scoring-Funktion** implementiert die Art, wie eine Aufgabe bewertet wird.
+Eine beispielhafte DSL-Implementierung könnte so aussehen:
+
+```
+fn score(single_choice_task t, string a) -> float {
```

Comment:
Dumme Frage: Da die Scoring-Funktion Bestandteil des Task-Structs ist: Muss ich da den Task übergeben? Sollte ich nicht einfach die korrekte Antwort übergeben?

Reply:
@malt-r Du hast im Beispiel mit `string` gearbeitet. Ist das dann Teil der Schnittstelle, d.h. müssen alle Scoring-Funktionen mit `string` arbeiten oder können die auch komplett anders aussehen? Gibt es irgendwo eine Art Template oder eine Basis-Signatur?

Reply:
> @malt-r Du hast im Beispiel mit `string` gearbeitet. Ist das dann Teil der Schnittstelle, d.h. müssen alle Scoring-Funktionen mit `string` arbeiten oder können die auch komplett anders aussehen? Gibt es irgendwo eine Art Template oder eine Basis-Signatur?

Das ist für mich noch nicht final entschieden, `string` wäre sicherlich ein Fall, mit dem wir alle Datentypen für Antworten irgendwie abbilden können. Aber zur Überprüfung von Zahlwerten ist das nicht optimal. Man könnte auch über ein `answer`-Datentyp nachdenken, der das irgendwie kapselt. Ich füge dafür mal eine Notiz hinzu.

Reply:
> Dumme Frage: Da die Scoring-Funktion Bestandteil des Task-Structs ist: Muss ich da den Task übergeben? Sollte ich nicht einfach die korrekte Antwort übergeben (wenn überhaupt, da ich in der Funktion ja im Task-Struct-Scope bin)?

Das würde sowas wie Member-Funktionen/Methoden implizieren, denen implizit immer das `task`-Objekt mit übergeben wird (analog z.B. zum `this`-Pointer). Das zu unterstützen ist aktuell nicht geplant. Ich dachte eher daran, dass mehrere `task`-Objekte eine Referenz auf die gleiche (unabhängig von den `task`-Objekten definierte) Scoring-Funktion bekommen können. 

Eine Alternative wäre, jeweils eine Standard-Implementierung für die Scoring-Funktion für jeden Task-Datentyp (`single_choice`, `multiple_choice`, etc.) fest vorzusehen, aber das liegt für mich konzeptionell schon gefährlich nah an OOP :)

Reply:
@malt-r Es wird dringend Zeit, dass Du Dein Konzept mal in der PL-Runde vorstellst und diskutierst. Solche Sachen kannst Du nicht "im stillen Kämmerlein" entscheiden - das ist keine Bachelorarbeit.

Reply:
@malt-r mach bitte ein/zwei tickets für die obige diskussion auf und resolve hier. ist ja erstmal nur doku.

Reply:
Tickets:
- #689
- #690